Modules/blake2module.c

```diff
@@ -0,0 +1,928 @@
+#define ECX_SSE4_2 (1 << 20)
+#define ECX_AVX (1 << 28)
+
+// EBX
+#define EBX_AVX2 (1 << 5)
+
+// EDX
+#define EDX_SSE (1 << 25)
+#define EDX_SSE2 (1 << 26)
+#define EDX_CMOV (1 << 15)
+
+// zero-initialized by default
+typedef struct {
+    bool sse, sse2, sse3, sse41, sse42, cmov, avx, avx2;
+    bool done;
+} cpu_flags;
+
+void detect_cpu_features(cpu_flags *flags) {
+  if (!flags->done) {
+    int eax1 = 0, ebx1 = 0, ecx1 = 0, edx1 = 0;
+    int eax7 = 0, ebx7 = 0, ecx7 = 0, edx7 = 0;
+#if defined(__x86_64__) && defined(__GNUC__)
+    __cpuid_count(1, 0, eax1, ebx1, ecx1, edx1);
+    __cpuid_count(7, 0, eax7, ebx7, ecx7, edx7);
+#elif defined(_M_X64)
+    int info1[4] = { 0 };
+    int info7[4] = { 0 };
+    __cpuidex(info1, 1, 0);
+    __cpuidex(info7, 7, 0);
+    eax1 = info1[0];
+    ebx1 = info1[1];
+    ecx1 = info1[2];
+    edx1 = info1[3];
+    eax7 = info7[0];
+    ebx7 = info7[1];
+    ecx7 = info7[2];
+    edx7 = info7[3];
+#else
+    (void) eax1; (void) ebx1; (void) ecx1; (void) edx1;
+    (void) eax7; (void) ebx7; (void) ecx7; (void) edx7;
+#endif
+
+    flags->avx = (ecx1 & ECX_AVX) != 0;
+
+    flags->avx2 = (ebx7 & EBX_AVX2) != 0;
+
+    flags->sse = (edx1 & EDX_SSE) != 0;
+    flags->sse2 = (edx1 & EDX_SSE2) != 0;
+    flags->cmov = (edx1 & EDX_CMOV) != 0;
+
+    flags->sse3 = (ecx1 & ECX_SSE3) != 0;
+    /* ssse3 = (ecx1 & ECX_SSSE3) != 0; */
+    flags->sse41 = (ecx1 & ECX_SSE4_1) != 0;
+    flags->sse42 = (ecx1 & ECX_SSE4_2) != 0;
+
+    flags->done = true;
+  }
+}
+
+static inline bool has_simd128(cpu_flags *flags) {
+  // For now this is Intel-only, could conceivably be #ifdef'd to something
+  // else.
+  return flags->sse && flags->sse2 && flags->sse3 && flags->sse41 && flags->sse42 && flags->cmov;
+}
+
+static inline bool has_simd256(cpu_flags *flags) {
+  return flags->avx && flags->avx2;
+}
+
+// Small mismatch between the variable names Python defines as part of configure
+// at the ones HACL* expects to be set in order to enable those headers.
+#define HACL_CAN_COMPILE_VEC128 HACL_CAN_COMPILE_SIMD128
+#define HACL_CAN_COMPILE_VEC256 HACL_CAN_COMPILE_SIMD256
+
+#include "_hacl/Hacl_Hash_Blake2b.h"
+#include "_hacl/Hacl_Hash_Blake2s.h"
+#if HACL_CAN_COMPILE_SIMD256
+#include "_hacl/Hacl_Hash_Blake2b_Simd256.h"
+#endif
+#if HACL_CAN_COMPILE_SIMD128
+#include "_hacl/Hacl_Hash_Blake2s_Simd128.h"
+#endif
+
+// MODULE TYPE SLOTS
+
+static PyType_Spec blake2b_type_spec;
+static PyType_Spec blake2s_type_spec;
+
+PyDoc_STRVAR(blake2mod__doc__,
+"_blake2b provides BLAKE2b for hashlib\n"
+);
+
+typedef struct {
+    PyTypeObject* blake2b_type;
+    PyTypeObject* blake2s_type;
+    cpu_flags flags;
+} Blake2State;
+
+static inline Blake2State*
+blake2_get_state(PyObject *module)
+{
+    void *state = PyModule_GetState(module);
+    assert(state != NULL);
+    return (Blake2State *)state;
+}
+
+static inline Blake2State*
+blake2_get_state_from_type(PyTypeObject *module)
+{
+    void *state = PyType_GetModuleState(module);
```

Comment:
```suggestion
    void *state = _PyType_GetModuleState(module);
```

You need to include `pycore_typeobject.h` for this.